fieldpath/set.go

```diff
@@ -455,3 +478,17 @@ func (s *SetNodeMap) iteratePrefix(prefix Path, f func(Path)) {
 		n.set.iteratePrefix(append(prefix, pe), f)
 	}
 }
+
+// Leaves returns a SetNodeMap containing
+// only setNodes with leaf PathElements.
+func (s *SetNodeMap) Leaves() *SetNodeMap {
+	out := &SetNodeMap{}
+	for _, n := range s.members {
+		next := setNode{
+			pathElement: n.pathElement,
+			set:         n.set.Leaves(),
+		}
+		out.members = append(out.members, next)
```

Comment:
Since we know how many members we have, I think we can pre-allocate the list beforehand. That probably has more impact that the double loop above :-)

Reply:
I've added preallocation both here and with the `PathElementSet` [here](https://github.com/kubernetes-sigs/structured-merge-diff/pull/185/files#diff-cbe655e4df83a27b44de60b4d4008cbcf550462c360f95d36aeb9826a2af525fR212)

makes it ever so slightly faster
```
BenchmarkFieldSet/leaves-20-16                            317458              3785 ns/op            2250 B/op         41 allocs/op
BenchmarkFieldSet/leaves-50-16                             82166             14462 ns/op            6881 B/op        141 allocs/op
BenchmarkFieldSet/leaves-100-16                            20643             60226 ns/op           29160 B/op        632 allocs/op
BenchmarkFieldSet/leaves-500-16                             2984            363610 ns/op          146550 B/op       3212 allocs/op
BenchmarkFieldSet/leaves-1000-16                            1407            851304 ns/op          325102 B/op       7189 allocs/op
```